### PR TITLE
mesa18: depends on python@3.8.12 max

### LINKED ISSUE #28219: mesa18@18.3.6 build fails: AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getchildren'
### Steps to reproduce the issue

`mesa18@18.3.6` build fails using:
* `spack@develop` (997dba8ffdc01d50dff3ebae791aff3d0145c818 from `Mon Jan 3 18:22:48 2022 +0100`)
* Ubuntu 20.04, x86_64
* GCC 9.3.0

Concrete spec: [mesa18-d7eiib.spec.yaml.txt](https://github.com/spack/spack/files/7804161/mesa18-d7eiib.spec.yaml.txt)

Concretization:
```
 -   mesa18@18.3.6%gcc@9.3.0+glx~llvm~opengles+osmesa patches=55a5611ca9fcbe8324c4d68a07b338134954ff12c5b122dc78ff376f012a1414 swr=none arch=linux-ubuntu20.04-x86_64
[+]      ^autoconf@2.69%gcc@9.3.0 patches=35c449281546376449766f92d49fc121ca50e330e60fefcfc9be2af3253082c2,7793209b33013dc0f81208718c68440c5aae80e7a1c4b8d336e382525af791a7,a49dd5bac3b62daa0ff688ab4d508d71dbd2f4f8d7e2a02321926346161bf3ee arch=linux-ubuntu20.04-x86_64
[+]          ^m4@1.4.19%gcc@9.3.0+sigsegv patches=9dc5fbd0d5cb1037ab1e6d0ecc74a30df218d0a94bdd5a02759a97f62daca573,bfdffa7c2eb01021d5849b36972c069693654ad826c1a20b53534009a4ec7a89 arch=linux-ubuntu20.04-x86_64
[+]              ^libsigsegv@2.13%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^perl@5.34.0%gcc@9.3.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
[+]              ^berkeley-db@18.1.40%gcc@9.3.0+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-ubuntu20.04-x86_64
[+]              ^bzip2@1.0.8%gcc@9.3.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
[+]                  ^diffutils@3.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]                      ^libiconv@1.16%gcc@9.3.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]              ^gdbm@1.19%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]                  ^readline@8.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]                      ^ncurses@6.2%gcc@9.3.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
[+]                          ^pkgconf@1.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]              ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-ubuntu20.04-x86_64
[+]      ^automake@1.16.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^binutils@2.36.1%gcc@9.3.0~gas+gold+headers~interwork+ld+libiberty~lto~nls+plugins libs=shared,static patches=a51b7bf4607645235be01683a3e2deb3a87d767cc124ea4ef6febf9f32abf4a7 arch=linux-ubuntu20.04-x86_64
[+]      ^bison@3.8.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^expat@2.4.1%gcc@9.3.0+libbsd arch=linux-ubuntu20.04-x86_64
[+]          ^libbsd@0.11.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]              ^libmd@1.0.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^flex@2.6.3%gcc@9.3.0+lex~nls arch=linux-ubuntu20.04-x86_64
[+]          ^findutils@4.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^gettext@0.21%gcc@9.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
[+]          ^libxml2@2.9.12%gcc@9.3.0~python arch=linux-ubuntu20.04-x86_64
[+]              ^xz@5.2.5%gcc@9.3.0+pic libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]          ^tar@1.34%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^glproto@1.4.17%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^util-macros@1.19.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^libtool@2.4.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^libx11@1.7.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^inputproto@2.3.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^kbproto@1.0.7%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^libxcb@1.14%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]              ^libpthread-stubs@0.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]              ^libxau@1.0.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]                  ^xproto@7.0.31%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]              ^libxdmcp@1.1.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]              ^xcb-proto@1.14.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^xextproto@7.3.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^xtrans@1.3.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^libxext@1.3.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]      ^py-mako@1.1.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]          ^py-markupsafe@2.0.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]              ^py-setuptools@59.4.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
[+]                  ^python@3.9.9%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87,4c2457325f2b608b1b6a2c63087df8c26e07db3e3d493caf36a56f0ecf6fb768,f2fd060afc4b4618fe8104c4c5d771f36dc55b1db5a4623785a4ea707ec72fb4 arch=linux-ubuntu20.04-x86_64
[+]                      ^libffi@3.3%gcc@9.3.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-ubuntu20.04-x86_64
[+]                      ^openssl@1.1.1l%gcc@9.3.0~docs certs=system arch=linux-ubuntu20.04-x86_64
[+]                      ^sqlite@3.36.0%gcc@9.3.0+column_metadata+fts~functions~rtree arch=linux-ubuntu20.04-x86_64
[+]                      ^util-linux-uuid@2.36.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
```

Error reproduced with entirely cached dependencies using:
* Docker container image = `ecpe4s/ubuntu20.04-runner-x86_64:2021-11-01`

```
$> spack mirror add E4S https://cache.e4s.io
$> spack buildcache keys -it

$> spack install --cache-only --include-build-deps --only dependencies -f ./mesa18-d7eiib.spec.yaml
OK

$> spack install --no-cache -f ./mesa18-d7eiib.spec.yaml
...
==> Installing mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q
==> Applied patch /spack/var/spack/repos/builtin/packages/mesa18/autotools-x11-nodri.patch
==> mesa18: Executing phase: 'autoreconf'
==> mesa18: Executing phase: 'configure'
==> mesa18: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

21 errors found in build log:
     13     libtoolize: copying file 'm4/ltsugar.m4'
     14     libtoolize: copying file 'm4/ltversion.m4'
     15     libtoolize: copying file 'm4/lt~obsolete.m4'
     16     autoreconf: running: /spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_
            path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholde/linux-ubuntu20.04-x86_64/gcc-9.3.0/autoconf-2.69-4wefch2dqfia5xr4oxseyr73xo7p7dav/bi
            n/autoconf --force
     17     autoreconf: configure.ac: not using Autoheader
     18     autoreconf: running: automake --add-missing --copy --force-missing
  >> 19     configure.ac:61: installing 'bin/ar-lib'
  >> 20     configure.ac:61: installing 'bin/compile'
  >> 21     configure.ac:45: installing 'bin/config.guess'
  >> 22     configure.ac:45: installing 'bin/config.sub'
  >> 23     configure.ac:46: installing 'bin/install-sh'
  >> 24     configure.ac:46: installing 'bin/missing'
     25     src/Makefile.am: installing 'bin/depcomp'
     26     parallel-tests: installing 'bin/test-driver'
     27     autoreconf: Leaving directory `.'
     28     ==> mesa18: Executing phase: 'configure'
     29     ==> [2022-01-03-13:40:09.388008] '/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/configure' '--prefix=/spack/opt/spack/[padded-to-256-cha
            rs]/linux-ubuntu20.04-x86_64/gcc-9.3.0/mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q' 'LDFLAGS=-L/spack/opt/spack/[padded-to-256-chars]/linux-ubuntu20.04-x86_64/gcc-9.3.0/ncurses-6
            .2-nxjjphalwbf2ywlxpn74nsfcsrq2bbtd/lib' '--enable-shared' '--disable-static' '--disable-libglvnd' '--disable-nine' '--disable-omx-bellagio' '--disable-omx-tizonia' '--disable-openc
            l' '--disable-opencl-icd' '--disable-va' '--disable-vdpau' '--disable-xa' '--disable-xvmc' '--disable-osmesa' '--with-vulkan-drivers=' '--disable-egl' '--disable-gbm' '--disable-dri
            ' '--enable-opengl' '--enable-gallium-osmesa' '--enable-glx=gallium-xlib' '--disable-gles1' '--disable-gles2' '--enable-shared-glapi' '--disable-llvm' '--with-platforms=x11' '--with
            -gallium-drivers=swrast' '--with-dri-drivers='
     30     checking build system type... x86_64-pc-linux-gnu

     ...

     670        self.process_element(file_name, doc)
     671      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./../../../../src/mapi/glapi/gen/gl_XML.py", line 914, in proce
            ss_element
     672        self.process_OpenGLAPI(file_name, element)
     673      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./../../../../src/mapi/glapi/gen/gl_XML.py", line 919, in proce
            ss_OpenGLAPI
     674        for child in element.getchildren():
     675    AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getchildren'
  >> 676    make[3]: *** [Makefile:939: ../../../../src/mesa/main/marshal_generated.c] Error 1
     677    make[3]: *** Waiting for unfinished jobs....
     678    Traceback (most recent call last):
     679      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./gl_table.py", line 207, in <module>
     680        main()
     681      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./gl_table.py", line 196, in main
     682        api = gl_XML.parse_GL_API(args.file_name)

     ...

     686        self.process_element(file_name, doc)
     687      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/gl_XML.py", line 914, in process_element
     688        self.process_OpenGLAPI(file_name, element)
     689      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/gl_XML.py", line 919, in process_OpenGLAPI
     690        for child in element.getchildren():
     691    AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getchildren'
  >> 692    make[3]: *** [Makefile:942: ../../../../src/mesa/main/marshal_generated.h] Error 1
     693    Traceback (most recent call last):
     694      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./gl_table.py", line 207, in <module>
     695    Traceback (most recent call last):
     696      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./gl_procs.py", line 190, in <module>
     697        main()
     698      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./gl_table.py", line 196, in main

     ...

     771        for child in element.getchildren():
     772    AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getchildren'
     773        self.process_OpenGLAPI(file_name, element)
     774      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/gl_XML.py", line 919, in process_OpenGLAPI
     775        for child in element.getchildren():
     776    AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getchildren'
  >> 777    make[3]: *** [Makefile:904: ../../../../src/mapi/glapi/glapi_mapi_tmp.h] Error 1
  >> 778    make[3]: *** [Makefile:945: ../../../../src/mesa/main/dispatch.h] Error 1
  >> 779    make[3]: *** [Makefile:914: ../../../../src/mapi/glapi/glapitable.h] Error 1
  >> 780    make[3]: *** [Makefile:948: ../../../../src/mesa/main/remap_helper.h] Error 1
  >> 781    make[3]: *** [Makefile:908: ../../../../src/mapi/glapi/glprocs.h] Error 1
  >> 782    make[3]: *** [Makefile:925: ../../../../src/mapi/glapi/glapi_x86-64.S] Error 1
  >> 783    make[3]: *** [Makefile:936: ../../../../src/mesa/main/api_exec.c] Error 1
  >> 784    make[3]: *** [Makefile:911: ../../../../src/mapi/glapi/glapitemp.h] Error 1
     785    Traceback (most recent call last):
     786      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./glX_proto_size.py", line 699, in <module>
     787        main()
     788      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/./glX_proto_size.py", line 693, in main
     789        api = gl_XML.parse_GL_API(args.filename, glX_XML.glx_item_factory())
     790      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/gl_XML.py", line 44, in parse_GL_API

     ...

     849        self.process_OpenGLAPI(file_name, element)
     850      File "/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen/gl_XML.py", line 919, in process_OpenGLAPI
     851        for child in element.getchildren():
     852    AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getchildren'
     853        for child in element.getchildren():
     854    AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getchildren'
  >> 855    make[3]: *** [Makefile:960: ../../../../src/glx/indirect_init.c] Error 1
  >> 856    make[3]: *** [Makefile:957: ../../../../src/glx/indirect.h] Error 1
     857    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src/mapi/glapi/gen'
  >> 858    make[2]: *** [Makefile:895: all-recursive] Error 1
     859    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src'
  >> 860    make[1]: *** [Makefile:680: all] Error 2
     861    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-mesa18-18.3.6-3kdtnixd32klhqamkwvzaveg5pj6ue5q/spack-src/src'
  >> 862    make: *** [Makefile:694: all-recursive] Error 1
```

### Information on your system

* **Spack:** 0.17.1-697-997dba8ffd
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-configure-args.txt](https://github.com/spack/spack/files/7804175/spack-configure-args.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/7804176/spack-build-env.txt)
[spack-build-02-configure-out.txt](https://github.com/spack/spack/files/7804177/spack-build-02-configure-out.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/7804171/spack-build-out.txt)

@ChristianTackeGSI 
@chuckatkins
@v-dobrev

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by eugeneswalker**
**Partially** fixes https://github.com/spack/spack/issues/28219

@renjithravindrankannath

### CODE REVIEW

**Review by quellyn — DISMISSED** [review]

**Comment by renjithravindrankannath** [comment]
Thank you @eugeneswalker. Build completed with python@3.8.12

**Comment by mwkrentel** [comment]
Although this patch helps, I'm concerned that it's incomplete.

I see the same problem with /usr/bin/python3 version 3.6.8 on RH 7.x.
And #28219 reports failure with python 3.8.10.  Both of these are <= 3.8.12,
so this patch wouldn't fix those cases.

I suspect that part of the problem is that mesa is missing a
dependency on some python module.  Maybe something like py-lxml, but
that's just a guess, I don't know what the right module is.

**Review by ChristianTackeGSI — DISMISSED** [review]
I haven't actually tried to understand the final problem, but generally, this doesn't look like it hurts too much.

**Comment by eugeneswalker** [comment]
> Although this patch helps, I'm concerned that it's incomplete.
> 
> And #28219 reports failure with python 3.8.10.

If you look at the concretization in https://github.com/spack/spack/issues/28219, it is actually building `mesa18` against `python@3.9.9`. The Python driving Spack is 3.8.10, but that isn't what mesa18 is using. 

For your case, you said you see the same problem with /usr/bin/python3 version 3.6.8 on RH 7.x, are you sure that is the python version being used to build Mesa18? Or, is it just the python version driving Spack?

**Comment by mwkrentel** [comment]
@eugeneswalker  This is going to turn into a long discussion.
Let's go back to #28219

**Comment by mwkrentel** [comment]
Ok, I withdraw my objection to this PR.

The python <= 3.8 issue is a small, separable issue.  The problem that
I reported is separate and requires a much larger fix.  So, you don't
want to hold this up for my problem.

Btw, my problem is that mesa18 usually won't build with an
external/system python (depending on some things like symlink).
If you build a new python, then it works, except for 3.9.

I suggest you change "fixes #28219" to "partially fixes" and leave
#28219 open for the other issue.

Btw, I also suggest changing `depends_on` to `python@:3.8` or
`python@:3.8.99` depending on your views on style.  This is an attempt
to predict the future.  If there is a python 3.8.13, it's more likely
than not to work, even if 3.9.x doesn't.

**Comment by eugeneswalker** [comment]
> I suggest you change "fixes #28219" to "partially fixes" and leave #28219 open for the other issue.

Done

> Btw, I also suggest changing `depends_on` to `python@:3.8` or `python@:3.8.99

Done

File: var/spack/repos/builtin/packages/mesa18/package.py [hunk 1]
```diff
@@ -33,7 +33,7 @@ class Mesa18(AutotoolsPackage):
     depends_on('flex', type='build')
     depends_on('gettext', type='build')
     depends_on('pkgconfig', type='build')
-    depends_on('python', type='build')
+    depends_on('python@:3.8.99', type='build')  # https://github.com/spack/spack/issues/28219
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa18/package.py]:
```suggestion
    depends_on('python@:3.8', type='build')  # https://github.com/spack/spack/issues/28219
```

**Review by chuckatkins — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
> I suspect that part of the problem is that mesa is missing a dependency on some python module. Maybe something like py-lxml, but that's just a guess, I don't know what the right module is.

Nah, the `xml` module is built into Python. It looks like the `getchildren` method was removed in Python 3.9, so this is the correct fix: https://docs.python.org/3.8/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.getchildren. A better fix might be to patch the library to use the suggested replacement method, but that can be done upstream or in a separate PR.

**Pull request merged**